RIOT-OS/RIOT
# dist/tools/kconfiglib: add configuration checks to genconfig.py

### Contribution description
This modifies the `genconfig.py` Kconfig script to perform checks on the configurations the user tried to set. It checks that all values that the user tried to set, were indeed set  (e.g. enabling modules or setting configuration parameters' values). This will catch invalid configurations like:
- trying to enable a module without its dependencies
- trying to set a configuration value out of its range
- trying to chose a `choice` option when its not available

This also introduces a convention for applications' Kconfig files: the `APPLICATION` symbol. This symbol is needed  in order to express dependencies on non-visible symbols (most likely features), and to use the `imply` attribute (weak reverse dependencies), which have a similar effect to the `FEATURES_OPTIONAL` but can work with any type of boolean symbol.

The convention is that, when the `APPLICATION` symbol is present (a non-visible boolean symbol), it should take the value `y` for the checks to pass. This allows, for example, to catch missing features like so:
```
config APLICATION
    bool
    default y
    depends on HAS_ARCH_32BIT
```

Enabling symbols when possible via this symbol looks like this:
```
config APLICATION
    bool
    default y
    imply MODULE_PERIPH_FLASHPAGE_RAW
```

All these checks are performed always, no only when `TEST_KCONFIG` is set, so it should be able to check invalid configurations on current applications only when using Kconfig.

If an invalid configuration is introduced, the build system will produce useful feedback on why the configuration is invalid. For example, this application works only on 8-bits platforms:
```
$ make all
[genconfig.py]:ERROR-=> The application symbol (APPLICATION) is not set.
[genconfig.py]:ERROR-   Check that the symbol defaults to 'y'.
[genconfig.py]:ERROR-   Check the following unmet dependencies: HAS_ARCH_8BIT (=n)
```

Tests for these checks are added in the `kconfiglib/tests` folder, which are included in the GitHub actions.

### Testing procedure
- Configurations via Kconfig should still work as usual.
- Run the test script `./dist/tools/kconfiglib/tests/test.sh`.
- You can try setting invalid configurations, they should be caught by the script.

### Issues/PRs references
Split from #14777
